### PR TITLE
[url] move url to library

### PR DESCRIPTION
**Pull request by bukepo**
This commit moves the URL parser to `src/lib`. Also includes the following changes:

* Split URL on initialization.
* Make a GetValue() const method.
* Check aLastValue.
* Add some negative tests.

### CODE REVIEW

File: src/posix/platform/hdlc_interface.cpp [hunk 1]
```diff
@@ -138,30 +138,30 @@ HdlcInterface::HdlcInterface(SpinelInterface::ReceiveFrameCallback aCallback,
 {
 }
 
-otError HdlcInterface::Init(Arguments &aArguments)
+otError HdlcInterface::Init(RadioUrl &aRadioUrl)
```
Comment (gjc13) [inline src/posix/platform/hdlc_interface.cpp:141]:
```suggestion
otError HdlcInterface::Init(const RadioUrl &aRadioUrl)
```

File: src/posix/platform/hdlc_interface.cpp [hunk 2]
```diff
@@ -409,7 +409,7 @@ otError HdlcInterface::WaitForWritable(void)
     return error;
 }
 
-int HdlcInterface::OpenFile(const char *aFile, Arguments &aArguments)
+int HdlcInterface::OpenFile(const char *aFile, RadioUrl &aRadioUrl)
```
Comment (gjc13) [inline src/posix/platform/hdlc_interface.cpp:412]:
```suggestion
int HdlcInterface::OpenFile(const char *aFile, const RadioUrl &aRadioUrl)
```

File: src/posix/platform/hdlc_interface.hpp [hunk 3]
```diff
@@ -74,14 +74,14 @@ class HdlcInterface
...
-     * @param[in]  aArguments         Arguments parsed from radio url.
+     * @param[in]  aRadioUrl         RadioUrl parsed from radio url.
      *
      * @retval OT_ERROR_NONE          The interface is initialized successfully
      * @retval OT_ERROR_ALREADY       The interface is already initialized.
      * @retval OT_ERROR_INVALID_ARGS  The UART device or executable cannot be found or failed to open/run.
      *
      */
-    otError Init(Arguments &aArguments);
+    otError Init(RadioUrl &aRadioUrl);
```
Comment (gjc13) [inline src/posix/platform/hdlc_interface.hpp:84]:
```suggestion
    otError Init(const RadioUrl &aRadioUrl);
```

File: src/posix/platform/hdlc_interface.hpp [hunk 4]
```diff
@@ -198,9 +198,9 @@ class HdlcInterface
     static void HandleHdlcFrame(void *aContext, otError aError);
     void        HandleHdlcFrame(otError aError);
 
-    static int OpenFile(const char *aFile, Arguments &aArguments);
+    static int OpenFile(const char *aFile, RadioUrl &aRadioUrl);
```
Comment (gjc13) [inline src/posix/platform/hdlc_interface.hpp:201]:
```suggestion
    static int OpenFile(const char *aFile, const RadioUrl &aRadioUrl);
```

File: src/posix/platform/spi_interface.cpp [hunk 5]
```diff
@@ -91,7 +91,7 @@ SpiInterface::SpiInterface(SpinelInterface::ReceiveFrameCallback aCallback,
 {
 }
 
-otError SpiInterface::Init(Arguments &aArguments)
+otError SpiInterface::Init(RadioUrl &aRadioUrl)
```
Comment (gjc13) [inline src/posix/platform/spi_interface.cpp:94]:
```suggestion
otError SpiInterface::Init(const RadioUrl &aRadioUrl)
```

File: src/posix/platform/spi_interface.hpp [hunk 6]
```diff
@@ -79,14 +79,14 @@ class SpiInterface
...
-     * @param[in]  aArguments         Arguments parsed from radio url.
+     * @param[in]  aRadioUrl         Arguments parsed from radio url.
      *
      * @retval OT_ERROR_NONE          The interface is initialized successfully.
      * @retval OT_ERROR_ALREADY       The interface is already initialized.
      * @retval OT_ERROR_INVALID_ARGS  The UART device or executable cannot be found or failed to open/run.
      *
      */
-    otError Init(Arguments &aArguments);
+    otError Init(RadioUrl &aRadioUrl);
```
Comment (gjc13) [inline src/posix/platform/spi_interface.hpp:89]:
```suggestion
    otError Init(const RadioUrl &aRadioUrl);
```

File: src/lib/url/url.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+#define SELF_TEST 0
+#endif
+
+#if SELF_TEST
+#include <assert.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Warray-bounds"
+
+void TestSimple()
```
Comment (jwhui) [inline src/lib/url/url.cpp:125]:
```suggestion
void TestSimple(void)
```

File: src/lib/url/url.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    assert(!strcmp(args.GetValue("baudrate"), "115200"));
+    assert(args.GetValue("not-exists") == nullptr);
+    assert(args.GetValue("last-value-wrong-position", url) == nullptr);
+    assert(args.GetValue("last-value-before-url", url - 1) == nullptr);
+    assert(args.GetValue("last-value-after-url", url + sizeof(url)) == nullptr);
+
+    printf("PASS %s\r\n", __func__);
+}
+
+void TestSimpleNoQueryString()
```
Comment (jwhui) [inline src/lib/url/url.cpp:142]:
```suggestion
void TestSimpleNoQueryString(void)
```

File: src/lib/url/url.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    assert(!args.Init(url));
+    assert(!strcmp(args.GetPath(), "/dev/ttyUSB0"));
+    assert(args.GetValue("last-value-wrong-position", url) == nullptr);
+    assert(args.GetValue("last-value-before-url", url - 1) == nullptr);
+    assert(args.GetValue("last-value-after-url", url + sizeof(url)) == nullptr);
+
+    printf("PASS %s\r\n", __func__);
+}
+
+void TestMultipleProtocols()
```
Comment (jwhui) [inline src/lib/url/url.cpp:156]:
```suggestion
void TestMultipleProtocols(void)
```

File: src/lib/url/url.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    ot::Url::Url args;
+
+    assert(!args.Init(url));
+    assert(!strcmp(args.GetPath(), "/dev/ttyUSB0"));
+    assert(!strcmp(args.GetValue("baudrate"), "115200"));
+
+    printf("PASS %s\r\n", __func__);
+}
+
+void TestMultipleProtocolsAndDuplicateParameters()
```
Comment (jwhui) [inline src/lib/url/url.cpp:168]:
```suggestion
void TestMultipleProtocolsAndDuplicateParameters(void)
```

File: src/lib/url/url.cpp [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    assert(args.GetValue("arg", url) == nullptr);
+    assert(args.GetValue("arg", url - 1) == nullptr);
+    assert(args.GetValue("arg", url + sizeof(url)) == nullptr);
+
+    printf("PASS %s\r\n", __func__);
+}
+
+#pragma GCC diagnostic pop
+
+int main()
```
Comment (jwhui) [inline src/lib/url/url.cpp:195]:
```suggestion
int main(void)
```

File: src/posix/platform/spi_interface.hpp [hunk 12]
```diff
@@ -79,14 +79,14 @@ class SpiInterface
      *
      * @note This method should be called before reading and sending spinel frames to the interface.
      *
-     * @param[in]  aArguments         Arguments parsed from radio url.
+     * @param[in]  aRadioUrl         Arguments parsed from radio url.
```
Comment (jwhui) [inline src/posix/platform/spi_interface.hpp:82]:
```suggestion
     * @param[in]  aRadioUrl          Arguments parsed from radio url.
```

File: src/posix/platform/hdlc_interface.hpp [hunk 13]
```diff
@@ -74,14 +74,14 @@ class HdlcInterface
      *
      * @note This method should be called before reading and sending spinel frames to the interface.
      *
-     * @param[in]  aArguments         Arguments parsed from radio url.
+     * @param[in]  aRadioUrl         RadioUrl parsed from radio url.
```
Comment (jwhui) [inline src/posix/platform/hdlc_interface.hpp:77]:
```suggestion
     * @param[in]  aRadioUrl          RadioUrl parsed from radio url.
```

File: src/posix/platform/radio_url.hpp [hunk 14]
```diff
@@ -34,47 +34,22 @@
...
-     * This method returns the url agument value.
-     *
-     * @param[in] aName       Argument name.
-     * @param[in] aLastValue  The last iterated argument value, nullptr for first value.
-     *
-     * @returns The argument value.
-     *
-     */
-    const char *GetValue(const char *aName, const char *aLastValue = nullptr);
+    RadioUrl(const char *aUrl);
```
Comment (abtink) [inline src/posix/platform/radio_url.hpp:45]: Adding documentation?

Comment (bukepo) [reply src/posix/platform/radio_url.hpp:55]: added

File: src/lib/url/url.hpp [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef OT_LIB_URL_URL_HPP_
+#define OT_LIB_URL_URL_HPP_
+
+#include <openthread/error.h>
+
+struct otUrl
```
Comment (jwhui) [inline src/lib/url/url.hpp:34]: nit: add doxygen for struct.

**Pull request merged**